src/shell/output.c

```diff
@@ -586,37 +587,73 @@ static int shell_output_parse_type (struct shell_output *out,
     return 0;
 }
 
-/* handle mustache templates and the similar special cases */
-static char *shell_output_get_path (struct shell_output *out, const char *path)
+static void shlog (void *arg,
+                   const char *file,
+                   int line,
+                   const char *func,
+                   const char *subsys,
+                   int level,
+                   const char *fmt,
+                   va_list ap)
 {
-    char *rv = NULL;
-    char *ptr;
+    char buf [4096];
+    int buflen = sizeof (buf);
+    int n = vsnprintf (buf, buflen, fmt, ap);
+    if (n >= buflen) {
+        buf[buflen-1] = '\0';
+        buf[buflen-2] = '+';
+    }
+    flux_shell_log (level, file, line, "%s", buf);
+}
 
-    /* replace {{id}} with jobid */
-    if ((ptr = strstr (path, "{{id}}"))) {
-        char buf[32];
-        int len, buflen, total_len;
 
-        len = strlen (path);
-        buflen = snprintf (buf, sizeof (buf), "%ju",
-                           (uintmax_t)out->shell->info->jobid);
-        /* -6 for {{id}}, +1 for NUL */
-        total_len = len - 6 + buflen + 1;
-        if (total_len > (PATH_MAX + 1)) {
-            errno = EOVERFLOW;
-            return NULL;
+static int mustache_cb (FILE *fp, const char *name, void *arg)
+{
+    flux_shell_t *shell = arg;
+    char value[128];
+
+    /*  "jobid" is a synonym for "id" */
+    if (strncmp (name, "jobid", 5) == 0)
+        name += 3;
+    if (strncmp (name, "id", 2) == 0) {
+        const char *type = "dec";
+        if (strlen (name) > 2) {
+            if (name[2] != '.') {
+                shell_log_error ("Unknown mustache tag '%s'", name);
+                return -1;
+            }
+            type = name+3;
+        }
+        if (flux_job_id_encode (shell->info->jobid,
+                                type,
+                                value,
+                                sizeof (value)) < 0) {
+            shell_log_errno ("flux_job_id_encode failed for %s", name);
+            return -1;
         }
-        if (!(rv = calloc (1, total_len)))
-            return NULL;
-        memcpy (rv, path, ptr - path);
-        memcpy (rv + (ptr - path), buf, buflen);
-        memcpy (rv + (ptr - path) + buflen, ptr + 6, len - (ptr - path) - 6);
     }
     else {
-        if (!(rv = strdup (path)))
-            return NULL;
+        shell_log_error ("Unknown mustache tag '%s'", name);
+        return -1;
+    }
+    return fputs (value, fp);
+}
+
+static char * shell_output_mustache_render (struct shell_output *out,
+                                            const char *path)
+{
+    struct mustache_renderer *mr;
+    char *result = NULL;
+
+    mr = mustache_renderer_create (mustache_cb, out->shell);
+    if (!mr) {
+        shell_log_errno ("mustache_renderer_create");
+        return NULL;
     }
-    return rv;
+    mustache_renderer_set_log (mr, (mustache_log_f) shlog, out->shell);
```

Comment:
`shlog()` doesn't use `out->shell`?

Reply:
Good catch thanks. `shlog()` was duplicated code, so I've now moved that function into `shell/log.c` and called it `shell_llog()` and removed the unnecessary `arg` parameter you caught.